src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -331,76 +335,59 @@ HeapWord* ParallelScavengeHeap::mem_allocate_work(size_t size,
       if (op.prologue_succeeded()) {
         assert(is_in_or_null(op.result()), "result not in heap");
 
-        // Exit the loop if the gc time limit has been exceeded.
-        // The allocation must have failed above ("result" guarding
-        // this path is null) and the most recent collection has exceeded the
-        // gc overhead limit (although enough may have been collected to
-        // satisfy the allocation).  Exit the loop so that an out-of-memory
-        // will be thrown (return a null ignoring the contents of
-        // op.result()),
-        // but clear gc_overhead_limit_exceeded so that the next collection
-        // starts with a clean slate (i.e., forgets about previous overhead
-        // excesses).  Fill op.result() with a filler object so that the
-        // heap remains parsable.
-        const bool limit_exceeded = size_policy()->gc_overhead_limit_exceeded();
-        const bool softrefs_clear = soft_ref_policy()->all_soft_refs_clear();
-
-        if (limit_exceeded && softrefs_clear) {
-          *gc_overhead_limit_was_exceeded = true;
-          size_policy()->set_gc_overhead_limit_exceeded(false);
-          log_trace(gc)("ParallelScavengeHeap::mem_allocate: return null because gc_overhead_limit_exceeded is set");
-          if (op.result() != nullptr) {
-            CollectedHeap::fill_with_object(op.result(), size);
-          }
-          return nullptr;
-        }
-
         return op.result();
       }
+      // Was the gc-overhead reached inside the safepoint? If so, this mutator should return null as well for global consistency.
+      if (_gc_overhead_counter >= GCOverheadLimitThreshold) {
+        return nullptr;
+      }
```

Comment:
Returning `nullptr` means the `OutOfMemoryException` should be thrown later. Is it good to add a `error` level log here?

Reply:
And notice: we can't identify the `OutOfMemoryException` whether because `gc overhead limit is exceeded`.

As I pointed out before: `the field MemAllocator::Allocation::_overhead_limit_exceeded are not used now`.

Reply:
The one inside the safepoint will print sth `log_info(gc)("GCOverheadLimitThreshold ...`. There can be multiple concurrent mutators reaching here; printing here is undesirable.

I don't think throwing OOM, from gc's perspective, is an "error".

Reply:
> I don't think throwing OOM, from gc's perspective, is an "error".

Nevermind; I just obey the statement in methods `MemAllocator::Allocation::check_out_of_memory` and `Universe::out_of_memory_error_java_heap`.

Reply:
Yes. I don't think it is exceeding overhead limit only problem, but OOM and heap dump have to be, in a sense of `atomic`. We encountered  this problem in real production system.

Reply:
> but OOM and heap dump have to be, in a sense of atomic

I see; it can indeed be annoying that heap-dump doesn't capture the intended snapshot -- a java thread can trigger a gc that changes heap-content. However, AFAICS, fixing that non-atomic issue requires updating callers of `report_java_out_of_memory`. Therefore, I am inclined to think it should be addressed in its own ticket. It's a preexisting issue on master as well, not sth newly introduced by this PR.

Reply:
Yes. Sounds good.